File: heapsort/heapsort.cpp
// Quick sort in C++

#include <iostream>
#include <sstream>
#include <fstream>
#include <vector>
#include <chrono>
#include <string> 
#include "../utils/load_data.h"
#include <typeinfo>
using namespace std;
using namespace std::chrono;

// function to swap elements
void swap(int *a, int *b) {
  int t = *a;
  *a = *b;
  *b = t;
}

void heapify(int arr[], int n, int root)
{
   int largest = root; 
   int l = 2*root + 1; 
   int r = 2*root + 2; 
  
   if (l < n && arr[l] > arr[largest])
   largest = l;
  
   if (r < n && arr[r] > arr[largest])
   largest = r;
  
   if (largest != root)
      {
      swap(arr[root], arr[largest]);
      heapify(arr, n, largest);
      }
}

void heapSort(int arr[], int n)
{

   for (int i = n / 2 - 1; i >= 0; i--)
   heapify(arr, n, i);
   for (int i=n-1; i>=0; i--)
   {
      swap(arr[0], arr[i]);
      heapify(arr, i, 0);
   }
}

// Driver code
int main() {
  int cantidades [15]= {100,1000,2000,3000,4000,5000,6000,7000,8000,9000,10000,20000,30000,40000,50000};
  for (size_t i = 0; i < 15; i++)
  {
    vector<int> tiempos;
    for (size_t j = 0; j < 5; j++)
    {
      vector<int> numeros;
      string nombreArchivo = "../data/"+to_string(cantidades[i])+".csv";
      numeros = readFileCsv(nombreArchivo);

      int longitud = numeros.size();
      int data[longitud];
      for (int i = 0; i < longitud; i++) {
          data[i] = numeros[i];
      }
      int n = sizeof(data) / sizeof(data[0]);
      // cout << "Unsorted Array: \n";
      // printArray(data, n);
      auto start = high_resolution_clock::now();
      heapSort(data, n);
      auto stop = high_resolution_clock::now();
      auto duration = duration_cast<nanoseconds>(stop - start);
      int tiempo = duration.count();
      tiempos.push_back(tiempo);
    }
    string nombreArchivo = "../output/heapsort/cpp/"+to_string(cantidades[i])+".csv";
    writeFileCsv(nombreArchivo, tiempos);
  }

}

File: heapsort/heapsort.py

import sys
sys.path.append('..\\')
import time
from utils.load_data import readFileCsv, writeFileCsv
# Heapsort in Python

def heapify(arr, n, i):
    largest = i  # Initialize largest as root
    l = 2 * i + 1  # left = 2*i + 1
    r = 2 * i + 2  # right = 2*i + 2
 
 # See if left child of root exists and is
 # greater than root
 
    if l < n and arr[i] < arr[l]:
        largest = l
 
 # See if right child of root exists and is
 # greater than root
 
    if r < n and arr[largest] < arr[r]:
        largest = r
 
 # Change root, if needed
 
    if largest != i:
        (arr[i], arr[largest]) = (arr[largest], arr[i])  # swap
 
  # Heapify the root.
 
        heapify(arr, n, largest)
 
 
# The main function to sort an array of given size
 
def heapSort(arr):
    n = len(arr)
 
 # Build a maxheap.
 # Since last parent will be at ((n//2)-1) we can start at that location.
 
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)
 
 # One by one extract elements
 
    for i in range(n - 1, 0, -1):
        (arr[i], arr[0]) = (arr[0], arr[i])  # swap
        heapify(arr, i, 0)

# data = [8, 7, 2, 1, 0, 9, 6]
cantidades = [100,1000,2000,3000,4000,5000,6000,7000,8000,9000,10000,20000,30000,40000,50000]

for x in cantidades:
  tiempos = []
  for y in range(6):
    data = readFileCsv('../data/'+str(x)+'.csv')
    size = len(data)
    start = time.time_ns()
    heapSort(data)
    end = time.time_ns()
    duration = end - start
    tiempos.append(duration)
  writeFileCsv('../output/heapsort/python/'+str(x)+'.csv', tiempos)
